src/autosave.js

```diff
@@ -0,0 +1,288 @@
+import PendingActions from '@ckeditor/ckeditor5-core/src/pendingactions';
+import DomEmitterMixin from '@ckeditor/ckeditor5-utils/src/dom/emittermixin';
+
+/* globals window */
+
+/**
+ * Autosave plugin provides an easy-to-use API to save the editor's content.
+ * It watches {@link module:engine/model/document~Document#event:change:data change:data}
+ * and `Window:beforeunload` events and calls the {@link module:autosave/autosave~SaveProvider#save} method.
+ *
+ * 		ClassicEditor
+ *			.create( document.querySelector( '#editor' ), {
+ *				plugins: [ ArticlePluginSet, Autosave ],
+ *				toolbar: [ 'heading', '|', 'bold', 'italic', 'link', 'bulletedList', 'numberedList', 'blockQuote', 'undo', 'redo' ],
+ *				image: {
+ *					toolbar: [ 'imageStyle:full', 'imageStyle:side', '|', 'imageTextAlternative' ],
+ *				}
+ *			} )
+ *			.then( editor => {
+ *				const autosave = editor.plugins.get( Autosave );
+ *				autosave.provider = {
+ *					save() {
+ *						const data = editor.getData();
+ *
+ *						// Note: saveEditorsContentToDatabase function might be async and return a promise to the saving action.
+ *						return saveEditorsContentToDatabase( data );
+ *					}
+ *				};
+ *			} );
+ */
+export default class Autosave extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'Autosave';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Provider is an object with the `save()` method. That method will be called whenever
+		 * the model's data changes. It might be called some time after the change,
+		 * since the event is throttled for performance reasons.
+		 *
+		 * @type {module:autosave/autosave~SaveProvider}
+		 */
+		this.provider = undefined;
+
+		/**
+		 * Throttled save method.
+		 *
+		 * @protected
+		 * @type {Function}
+		 */
+		this._throttledSave = throttle( this._save.bind( this ), 500 );
+
+		/**
+		 * Last document version.
+		 *
+		 * @protected
+		 * @type {Number}
+		 */
+		this._lastDocumentVersion = editor.model.document.version;
+
+		/**
+		 * @private
+		 * @type {DomEmitterMixin}
+		 */
+		this._domEmitter = Object.create( DomEmitterMixin );
+
+		/**
+		 * Save action counter monitors number of actions.
+		 *
+		 * @private
+		 * @type {Number}
+		 */
+		this._saveActionCounter = 0;
+
+		/**
+		 * An action that will be added to pending action manager for actions happening in that plugin.
+		 *
+		 * @private
+		 * @type {Object|null}
+		 */
+		this._action = null;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const doc = editor.model.document;
+		const pendingActions = editor.plugins.get( PendingActions );
+
+		this.listenTo( doc, 'change:data', () => {
+			this._incrementCounter();
+
+			const willOriginalFunctionBeCalled = this._throttledSave();
+
+			if ( !willOriginalFunctionBeCalled ) {
+				this._decrementCounter();
```

Comment:
Why do you increment counter first and then decrease it if `willOriginalFunctionBeCalled` is false? Can't you just increment the counter if willOriginalFunctionBeCalled is `true`?

Reply:
Ach, I see that `_throttledSave` can synchronously execute an action, so it makes sense.

Reply:
On the other hand, I believe that it would be better if `throttle` allways call the function asynchronously, so if it is the first action it is called with timeout `0`.

Reply:
I'll check the `lodash` implementation and will align to it.

Reply:
*lodash* fires the first call instantly after the wrapped fn is being called, so IMO we can also call the fn instantly.